types/universalify/index.d.ts

```diff
@@ -1,7 +1,21 @@
-// Type definitions for universalify 1.0
+// Type definitions for universalify 2.0
 // Project: https://github.com/RyanZim/universalify#readme
-// Definitions by: Richie Bendall <https://github.com/Richienb>
+// Definitions by: Richie Bendall <https://github.com/Richienb>, Jan Peer Stöcklmair <https://github.com/JPeer264>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 3.7
 
-export function fromCallback(fn: (...args: any[]) => any): (...args: any[]) => Promise<any> | void;
-export function fromPromise(fn: (...args: any[]) => any): (...args: any[]) => Promise<any> | void;
+import { List } from 'ts-toolbelt';
+
+export interface UniversalifyResult<T, A extends any[]> {
+    (...args: List.Append<A, (err: any, result?: T) => void>): void;
+    (...arg1: A): Promise<T>;
+}
+
+/**
+ * Returns promise if the last argument is not a callback
+ * Returns void and puts the result in the callback, which is the last argument of the function
+ */
+export function fromCallback<T, A extends any[]>(
+    func: (...args: List.Append<A, (err: any, result?: T) => void>) => void,
```

Comment:
It's better to just inline the [types](https://github.com/millsp/ts-toolbelt/blob/94ae2f12be155e8d9efc7328116016384910b271/src/List/Append.ts#L12):

```suggestion
    func: (...args: [...A, (err: any, result?: T) => void]) => void,
```

Reply:
Unfortunately this is a TypeScript v4 feature. As here the spread operator is at the beginning of the arguments list. This is why it needs to have multiple generics for older TypeScript versions than v4.0

Reply:
ref: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46800#issuecomment-678043285